webpush/encrypt.go

```diff
@@ -238,14 +293,48 @@ func newContext(clientPublicKey, serverPublicKey []byte) []byte {
 // The context argument should match what newContext creates
 func newInfo(infoType string, context []byte) []byte {
 	var info []byte
-	info = append(info, []byte("Content-Encoding: ")...)
+	info = append([]byte{}, []byte("Content-Encoding: ")...)
 	info = append(info, []byte(infoType)...)
 	info = append(info, 0)
-	info = append(info, []byte("P-256")...)
-	info = append(info, context...)
+	// For aes128gcm ctx is not needed.
+	if len(context) > 0 {
+		info = append(info, []byte("P-256")...)
+		info = append(info, context...)
+	}
+	return info
+}
+
+// Returns a keying material. See sections of
+// https://tools.ietf.org/html/draft-ietf-webpush-encryption-07#section-3.3
+func newKeyInfo(subscriptionPublicKey, serverPublicKey []byte) []byte {
+	info := make([]byte, len(keyInfoPrefix)+len(subscriptionPublicKey)+len(serverPublicKey))
+	n := copy(info, keyInfoPrefix)
+	n += copy(info[n:], subscriptionPublicKey)
+	copy(info[n:], serverPublicKey)
 	return info
 }
 
+// Returns an 86 octet header. See section 2.1 of:
+// https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-06#section-2.1
+func appendHeader(salt, serverPublicKey, ciphertext []byte) []byte {
+	var result []byte
+	// A push service is not required to support more than 4096 octets of
+	// payload body so the record size can be at most 4096.
+	rs := uint32(maxPayloadRecordSize)
+	rsbuf := make([]byte, 4)
+	binary.BigEndian.PutUint32(rsbuf, rs)
```

Comment:
no need for `rs` really, just do `PutUint32(rsbuf, uint32(maxPayloadRecordSize))`